packages/Search/src/DTK_DistributedSearchTree_decl.hpp

```diff
@@ -0,0 +1,118 @@
+/****************************************************************************
+ * Copyright (c) 2012-2017 by the DataTransferKit authors                   *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the DataTransferKit library. DataTransferKit is     *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ ****************************************************************************/
+
+#ifndef DTK_DISTRIBUTED_SEARCH_TREE_DECL_HPP
+#define DTK_DISTRIBUTED_SEARCH_TREE_DECL_HPP
+
+#include <Kokkos_View.hpp>
+
+#include <Teuchos_Comm.hpp>
+#include <Teuchos_RCP.hpp>
+
+#include <DTK_DBC.hpp>
+#include <DTK_LinearBVH.hpp>
+#include <details/DTK_DetailsDistributedSearchTreeImpl.hpp>
+
+#include "DTK_ConfigDefs.hpp"
+
+namespace DataTransferKit
+{
+
+template <typename DeviceType>
+class DistributedSearchTree
+{
+  public:
+    DistributedSearchTree(
+        Teuchos::RCP<Teuchos::Comm<int> const> comm,
+        Kokkos::View<Box const *, DeviceType> bounding_boxes );
+
+    /** \brief Finds object meeting the passed predicates (e.g. nearest to some
+     *  point or overlaping with some box)
+     *
+     *  This query function performs a batch of spatial or k-nearest neighbors
+     * searches.
+     *  The results give indices of the objects that meet predicates (as given
+     *  to the constructor).  They are organized in a distributed compressed
+     *  row storage format.
+     *
+     *  \c indices stores the indices of the objects that met the predicates.
+     *  \c offset stores the locations in the \c indices view that start a
+     *  predicate, that is, \c queries( q ) is met by \indices( o ) for \c
+     *  objects( q ) <= o < objects( q + 1 ) that live on processes \ranks( o )
+     *  respectively.  Following the usual convention, \c offset( n ) = nnz,
+     *  where \c n is the number of queries that were performed and \c nnz is
+     *  the total number of collisions.
+     *
+     *  \note The views \c indices, \c offset, and \c ranks are passed by
+     *  reference because Kokkos::realloc() calls the assignment operator.
+     *
+     *  \param[in] queries Collection of predicates of the same type.  These
+     *  may be spatial predicates or nearest predicates.
+     *  \param[out] indices Object local indices that met the predicates.
+     *  \param[out] offset Array of predicate offsets for one-dimensional
```

Comment:
Just a little bit more about the layout of the output arrays would be great. Are there multiple indices returned per query?